packages/utils/telemetry-utils/src/errorLoggingInternalHelpers.ts

```diff
@@ -0,0 +1,94 @@
+};
+
+/** Inspect the given error for common "safe" props and return them */
+export function extractLogSafeErrorProperties(error: any) {
+    const removeMessageFromStack = (stack: string, errorName?: string) => {
+        const stackFrames = stack.split("\n");
+        stackFrames.shift(); // Remove "[ErrorName]: [ErrorMessage]"
+        if (errorName !== undefined) {
+            stackFrames.unshift(errorName); // Add "[ErrorName]"
+        }
+        return stackFrames.join("\n");
+    };
+
+    const message = (typeof error?.message === "string")
+        ? error.message as string
+        : String(error);
+
+    const safeProps: { message: string; errorType?: string; stack?: string } = {
+        message,
+    };
+
+    if (isRegularObject(error)) {
+        const { errorType, stack, name } = error;
+
+        if (typeof errorType === "string") {
+            safeProps.errorType = errorType;
+        }
+
+        if (typeof stack === "string") {
+            const errorName = (typeof name === "string") ? name : undefined;
+            safeProps.stack = removeMessageFromStack(stack, errorName);
+        }
+    }
+
+    return safeProps;
+}
+
+/** Copy props from source onto target, overwriting any keys that are already set on target */
+export function copyProps(target: unknown, source: ITelemetryProperties) {
+    Object.assign(target, source);
+}
+
+/**
+ * Read-Write Logging Error.  Not exported.
+ * This type alias includes addTelemetryProperties, and applies to objects even if not instanceof LoggingError\
+ */
+type RwLoggingError = LoggingError;
+
+/** Type guard for RwLoggingError.  Not exported. */
+const isRwLoggingError = (x: any): x is RwLoggingError =>
+    typeof x?.addTelemetryProperties === "function" && isILoggingError(x);
+
+/**
+ * Mixes in the given properties to the given error object, then accessible via ILoggingError.getTelemetryProperties
+ * @param errorObject - An object (MUST be non-null non-array object type, not frozen) to add telemetry features to
+ * @param props - The telemetry props to add to the errorObject
+ * @returns the same object that was passed in, now also implementing ILoggingError.
+ */
+export function mixinTelemetryProps<T>(
+    errorObject: T,
+    props: ITelemetryProperties,
+): asserts errorObject is T & ILoggingError {
+    assert(isRegularObject(errorObject) && !Object.isFrozen(errorObject), "Cannot mixin Telemetry Props");
+
+    if (isRwLoggingError(errorObject)) {
+        errorObject.addTelemetryProperties(props);
+        return;
+    }
+
+    // Even though it's not exposed, fully implement RwLoggingError for subsequent calls to mixinTelemetryProps
+    const loggingError = errorObject as T & RwLoggingError;
+
+    const propsForError = {...props};
+    loggingError.getTelemetryProperties = () => propsForError;
+    loggingError.addTelemetryProperties =
+        (newProps: ITelemetryProperties) => { copyProps(propsForError, newProps); };
```

Comment:
I'm not sure it's good idea to patch objects that way, or it's needed.
I think callers should differentiate much earlier if this error object is isILoggingError() or not. And if not - build error object from scratch, copying properties as needed. I.e. same simple message - normalize if it's of wrong shape, or annotate if we know for sure it's isILoggingError(). I do not think we should support or consider some states in between (i.e. object, not frozen, maybe has error code, maybe isILoggingError). B/W picture is just easier to work with - much less cases to support test, much simpler guidance to developers, etc.

Reply:
I added this function in PR #6748, in order to reduce dependency on `LoggingError`.  In this PR I'm moving it to a non-imported file -- devs should use `normalizeError` or `annotateErrorObject`, somewhat to your point.

Taking a step back,  I disagree with this:

>  I do not think we should support or consider some states in between (i.e. object, not frozen, maybe has error code, maybe isILoggingError)

 We just can't really rely on errors being well-typed throughout, because JS always presents errors to `catch` as `any`. i.e. there is no B/W picture available. I'm really committed to enabling devs to work with any arbitrary error that shows up.

As for this:

> I think callers should differentiate much earlier if this error object is isILoggingError() or not.

Why should they care?  The logging code itself handles `any` input, checking for the interface and if present using it.  I want to simplify the dev story to only always point folks to `normalizeError` and `annotateErrorObject`.  Really, `LoggingError`  and even the interface can be implementation details.

Reply:
It escapes me why it needs to be complicated (you might have some scenarios in mind, if you do, please share).
Yes, we can get random thing in catch(), but all we need is normalizeError aka wrap() that does something reasonable, and the rest of the code just works with LoggingError & IFluidErrorBase.
I'd just merge IFluidErrorBase requirement into isILoggingError (as we are basically installing new requirement - all code should provide errorCode & errorType) and have one type to work with - I think complexity would go down substantially. Basically either code works with well-known type and can call methods on it and read/write properties, or it has no clue and it relies on normalizeError() to give it that type. I do not see that much need yet for things in the middle (i.e. IFluidErrorBase but not isILoggingError, isILoggingError but not IFluidErrorBase).

And it's easy to keep adding new requirements without adding complexity (i.e. if we add object freezing, then we can route code to take normalizeError() path that is guaranteed to return r/w object of known type - all the complexity stays in a exactly two places - test function, and normalization function, they need to be in sync, but otherwise all the code around them does not care about any of these details).

Reply:
Here's the difference we need to decide about:  Is there a value to keeping the same error object itself and annotating it as-is, or is it just as good to quietly wrap it?

I'll address this more in a top-level comment since it's so central, but I find it _less_ complicated (from a usage POV) to mixin/annotate the original error rather than `normalizeError` continually wrapping and re-wrapping.

Reply:
i don't think there is much value in preserving the javascript object as long as it's properties, and most importantly stack are preserved if appropirate.

Reply:
@vladsud - Discussed offline, I believe we are ok with this approach, right?  If not, please open an issue since it's already checked in this way (this is mostly a file move)